### PR TITLE
Wait for prepare hooks to run and collect changed files in the same livesync run

### PR DESCRIPTION
**Pull request by hdeshev**
- Allow for calling out to compilers in hooks and collecting changed files
afterwards.
- Allow changes to excluded files (*.ts) to trigger a sync, and filter
those files out after the platform is prepared.
- Fix livesync file filter bug on ios.

### CODE REVIEW

**Comment by justcodebuilduser** [comment]
:heart:

File: services/usb-livesync-service-base.ts [hunk 1]
```diff
@@ -20,32 +20,49 @@ class SyncBatch {
...
+			if (this.syncQueue.length > 0) {
+				this.$logger.trace("Syncing %s", this.syncQueue.join(", "));
+				this.$dispatcher.dispatch( () => {
+					return (() => {
+						this.done();
+					}).future<void>()();
+				});
+			}
+			this.timer = null;
+		}, 250); // https://github.com/Microsoft/TypeScript/blob/master/src/compiler/tsc.ts#L487-L489
```
Comment (teobugslayer) [inline services/usb-livesync-service-base.ts]: Consider introducing a named constant, please

File: services/usb-livesync-service-base.ts [hunk 2]
```diff
@@ -277,21 +297,27 @@ export class UsbLiveSyncServiceBase implements IUsbLiveSyncServiceBase {
...
-					this.$logger, this.$dispatcher, (filesToSync) => {
+					this.$logger, this.$dispatcher, data, () => {
 						this.preparePlatformForSync(data.platform);
-						this.$iOSEmulatorServices.syncFiles(data.appIdentifier, data.projectFilesPath, filesToSync, data.notRunningiOSSimulatorAction,  data.getApplicationPathForiOSSimulatorAction, data.iOSSimulatorRelativeToProjectBasePathAction);
+
+						setTimeout(() => {
+							this.batch.syncFiles((filesToSync) => {
+								this.$iOSEmulatorServices.syncFiles(data.appIdentifier, data.projectFilesPath, filesToSync, data.notRunningiOSSimulatorAction,  data.getApplicationPathForiOSSimulatorAction, data.iOSSimulatorRelativeToProjectBasePathAction);
+							});
+						}, 250);
```
Comment (teobugslayer) [inline services/usb-livesync-service-base.ts]: The same named constant may be used here

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

File: services/usb-livesync-service-base.ts (hunk 1 above)
Comment (hdeshev) [inline services/usb-livesync-service-base.ts]: Fixed.

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by teobugslayer** [comment]
We are postponing this change for after 1.5.2 release to have additional time to properly test it. It is still targeting RELEASE branch though.

**Comment by teobugslayer** [comment]
:+1: squash and merge after we release 1.5.2

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by Mitko-Kerezov** [comment]
Closing this and will retarget to `master`

**Pull request closed without merging**